src/layers/core/geojson-layer/geojson-layer.js

```diff
@@ -155,64 +120,116 @@ export default class GeoJsonLayer extends Layer {
-        onHover,
-        onClick
+    const {subLayers: {pointFeatures, lineFeatures, polygonFeatures}} = this.state;
+    const {id, getStrokeColor, getStrokeWidth, getFillColor, getHeight} = this.props;
+    const {extruded, wireframe} = this.props;
+
+    let {drawPoints, drawLines, drawPolygons, fillPolygons} = this.props;
+    drawPoints = drawPoints && pointFeatures && pointFeatures.length > 0;
+    drawLines = drawLines && lineFeatures && lineFeatures.length > 0;
+    drawPolygons = drawPolygons && polygonFeatures && polygonFeatures.length > 0;
+    fillPolygons = fillPolygons && polygonFeatures && polygonFeatures.length > 0;
+
+    // Filled Polygon Layer
+    const polygonFillLayer = fillPolygons && new PolygonLayer(Object.assign({}, this.props, {
+      id: `${id}-polygon-fill`,
+      data: polygonFeatures,
+      getPolygon: f => Container.get(f, 'geometry.coordinates'),
+      getHeight,
+      getColor: getFillColor,
+      extruded,
+      wireframe: false,
+      onHover: this._onHoverPolygon,
+      onClick: this._onClickPolygon
+    }));
+
+    // Polygon outline or wireframe
+    let polygonOutlineLayer = null;
+    if (drawPolygons && extruded && wireframe) {
+      polygonOutlineLayer = new PolygonLayer(Object.assign({}, this.props, {
+        id: `${id}-polygon-wireframe`,
+        data: polygonFeatures,
+        getPolygon: f => Container.get(f, 'geometry.coordinates'),
+        getHeight,
+        getColor: getFillColor,
+        extruded: true,
+        wireframe: true,
+        onHover: this._onHoverPolygon,
+        onClick: this._onClickPolygon
       }));
-
-    const pathLayer = pathLayerData &&
-      new PathLayer(Object.assign({}, this.props, {
-        id: `${id}-path-sublayer`,
-        data: pathLayerData,
-        getPaths: this._extractPaths,
+    } else if (drawPolygons) {
+      polygonOutlineLayer = new PathLayer(Object.assign({}, this.props, {
+        id: `${id}-polygon-outline`,
+        data: polygonFeatures,
+        getPaths: f => Container.get(f, 'geometry.coordinates'),
         getColor: getStrokeColor,
         getWidth: getStrokeWidth,
-        getOpacity: getStrokeOpacity,
-        pickable: false,
-        onHover,
-        onClick
-      }));
-
-    const polygonLayer = polygonLayerData &&
-      new PolygonLayer(Object.assign({}, this.props, {
-        id: `${id}-polygon-sublayer`,
-        data: polygonLayerData,
-        getPolygons: this.getFeaturePolygons,
-        getColor: getFillColor,
-        getOpacity: getFillOpacity,
-        pickable: false,
-        onHover,
-        onClick
+        onHover: this._onHoverPolygon,
+        onClick: this._onClickPolygon
       }));
+    }
 
-    return [polygonLayer, pathLayer, pointLayer].filter(Boolean);
+    const lineLayer = drawPolygons && drawLines && new PathLayer(Object.assign({}, this.props, {
+      id: `${id}-line-paths`,
+      data: lineFeatures,
+      getPaths: f => Container.get(f, 'geometry.coordinates'),
```

Comment:
I agree we should try to avoid unnecessary data transform when possible, my hunch is that the performance loss due to array repacking won't be as salient (what was the absolute delta time for the 5x speedup in the bench test?)

Regarding picking, I agree at the geoJson layer level, we should keep the original feature, but it seems strange that the general polygon and path layer need to understand geoJson and surface a geoJson feature when hovering. I will think about it too and come back in the later comment.